src/hpc/tls.h

```diff
@@ -55,28 +98,96 @@ typedef struct ThreadLocalStorage
   UInt LocksContended;
 } ThreadLocalStorage;
 
-#ifdef HAVE_NATIVE_TLS
+#ifdef USE_NATIVE_TLS
 extern __thread ThreadLocalStorage *TLSInstance;
 #endif
 
+#ifdef USE_PTHREAD_TLS
+
+#include <pthread.h>
+
+#ifdef ALLOW_PURE_PTHREAD_GETSPECIFIC
+// pthread_getspecific() is not defined as pure by default; redeclaring
+// it as such works for gcc/clang and allows the compiler to hoist calls
+// to pthread_getspecific() out of loops and perform constant
+// subexpression elimination on expressions containing TLS values.
+PURE_FUNC void * pthread_getspecific(pthread_key_t);
+#endif
+
+#ifdef USE_MACOS_PTHREAD_TLS_ASM
+static UInt TLSOffset;
+UInt GetTLSOffset(void);
+#else
+static pthread_key_t TLSKeyCopy;
+pthread_key_t GetTLSKey(void);
+#endif
+
+// This constructor function will initialize the static variable for
+// each C module. Making the variable static allows optimizations that
+// a global non-static variable would not permit. On some architectures,
+// it is also faster to access a static than a global non-static
+// variable, avoiding one level of indirection.
+//
+// Constructor functions are executed upon program start/library load.
+#ifdef USE_MACOS_PTHREAD_TLS_ASM
+CONSTRUCTOR_FUNC static void InitTLSOffset()
+{
+    TLSOffset = GetTLSOffset();
+}
+#elif defined(HAVE_FUNC_ATTRIBUTE_CONSTRUCTOR)
+CONSTRUCTOR_FUNC static void InitTLSKey()
```

Comment:
You already used `CONSTRUCTOR_FUNC` above, so I think the `#ifdef` logic needs to be changed a bit:
```C
#ifdef HAVE_FUNC_ATTRIBUTE_CONSTRUCTOR
  #ifdef USE_MACOS_PTHREAD_TLS_ASM
  ...
  #else
  ...
  #endif
#endif
```

Reply:
Good catch. This is actually a problem with code that can only possibly be compiled on non-gcc/clang compilers in that it's difficult to test for that eventuality (especially as icc and others also include the gcc/clang extensions).